vllm/v1/attention/backends/flashinfer.py

```diff
@@ -218,22 +218,46 @@ def __post_init__(self):
 
 
 class FlashInferMetadataBuilder(AttentionMetadataBuilder[FlashInferMetadata]):
+    full_cudagraph_supported: ClassVar[bool] = True
+    force_separate_routine: ClassVar[Optional[bool]] = True
 
     def __init__(self, runner: GPUModelRunner, kv_cache_spec: AttentionSpec,
                  block_table: BlockTable):
         self.runner = runner
+        self.vllm_config = runner.vllm_config
         self._workspace_buffer = None
         self._prefill_wrapper = None  # Wrapper for prefill/append
-        self._decode_wrapper = None  # Wrapper for decode
+        self._decode_wrapper = None  # Wrapper for decode (general shape)
+        self.enable_cuda_graph = (
+            self.vllm_config.compilation_config.full_cuda_graph)
+        if self.enable_cuda_graph:
+            # For full cudagraph capture, one `decode_wrapper` for each batch
+            # size is needed for FlashInfer.
+            self._decode_wrappers_cudagraph: dict[
+                int, BatchDecodeWithPagedKVCacheWrapper] = {}
+            self._decode_cudagraph_max_bs = min(
+                runner.max_num_reqs, runner.cudagraph_batch_sizes[-1])
```

Comment:
Even if `max_num_reqs` is larger than `cudagraph_batch_sizes[-1]`, that batch would not run with cudagraphs, so why not just do this:
```suggestion
            self._decode_cudagraph_max_bs = runner.cudagraph_batch_sizes[-1]
```

Reply:
This is aligned with the logic in the `capture_model` in gpu_model_runner.py
>decode_cudagraph_batch_sizes = [
                    x for x in self.cudagraph_batch_sizes if x <= max_num_reqs]

where the max batch size for decode cudagraph capturing is min(max_num_reqs, cudagraph_batch_sizes[-1])).  

The  `_decode_cudagraph_max_bs` is to strictly distinguish whether a decode batch run in cudagraph or not. So we can safely select a decode wrapper for cudagraph support from `self._decode_wrappers_cudagraph` (dict) that has been captured. or fall back to `self._decode_wrapper` for the general shape (bs >`_decode_cudagraph_max_bs`).